### ISSUE TITLE
LOD Updates in Cruise Phase

### ISSUE THREAD

**Issue by ResetXPDR** [Fragtality/DynamicLOD#13]
I and a few others over at AVSIM [here](https://www.avsim.com/forums/topic/640425-msfs-adaptive-lod/page/34/) have observed that the LOD values stay the same when in cruise, which is fine if you are way up high in an airliner, but not so much when cruising lower over undulating territory like in the elevation profile attached.
![DynamicLOD cruise altitude profile](https://github.com/Fragtality/DynamicLOD/assets/156568588/d06eae13-3388-4141-8322-c6de106215d2)

I have experience with coding over the last 40 years, but am new to the github form of collaborative development. Nonetheless, I managed to download your source code, find the area where LODs are chosen and modified the code so that in cruise it will change LOD but only if it exceeds a hysteresis threshold for the next AGL bracket. I managed to compile it and get it to work with great results on my system but am unsure how to submit the suggested code changes, so I will just attach them here (had to put a txt extension on the file otherwise it wouldn't upload).
[LODController.cs.txt](https://github.com/Fragtality/DynamicLOD/files/13933097/LODController.cs.txt)

In short, I added two extra else if cases at line 152 of LODController.cs that don't care about VS trend but will only trigger once the hysteresis threshold has been breached, which I have currently set to +/-5% but this could always be user selectable in the GUI if you want to spruce it up a bit. The original climb and descent logic remains intact and that is what is executed first if applicable.

**Comment by Fragtality** [comment]
Ok I put aside for the Moment that I stated not being interested in Suggestions and that you have reported that Issue in the wrong Repo first.

Then still: I don't see any Benefits and only Drawbacks. The whole Idea to consider the VS (SimVar `VERTICAL SPEED`) is to *NOT* change the LOD just because the Terrain climbs and descends while the Plane does not. So ignoring that just leads to Situations of constantly changed LODs during Cruise.
And generally I just can't follow why hysteresis should be applied here at all?! There is no delay between Action and Cause . I'm not an Engineer ofc and only now that from Thermostats ... but: That type of Oscillation is exactly what was meant to be excluded by that.

Sure, there are Corner-Cases like shown in your Flight Profile. Descending "so near to the Ground" but still high above your Destination will reduce LODs earlier than it would be needed. BUT: The main Goal of having a reduced LOD on Ground (on your Destination) is given and each Decrement is exactly executed *once* and Increments are *excluded*.

I'm aware of Cases where DynamicLOD somehow skips Pairs respectively did not increase to the intended LOD for Cruise where I have also a Log for that. Have to look into this further, it almost seems as the ref Keyword would not work - which can't really be. It finds a higher Pair, still does not want to change to that. Maybe I just add a special Case for Cruise (zero Trend) or something like that.
When I have Time for that. Don't see it really as urgent Issue, since using a lower LOD on while Ground is still the Case.

**Comment by ResetXPDR** [comment]
Perhaps null zone is a better term to use than hysteresis in this case. The reason for its inclusion is precisely to NOT have LOD changing in cruise with only small AGL deviations. I've done a number of test flights with the change and it does not constantly change LOD like you suggest it would. The original climb/descent logic remains the same, it is only in cruise that it changes and even then it is only when cruising at low level as the null zone is quite high at high altitude and the LODs stay the same. 

As such, IMO, and it seems a few others over at AVSIM, it makes it better for low level IFR and has no effect on your original objective of reducing LOD at airports on the ground or while climbing or descending. Perhaps you should try it before dismissing it entirely. I hope you do see the benefit and incorporate this change.

**Comment by ResetXPDR** [comment]
I will also add that the change I made also fixes an issue where you change profile mid flight while in cruise and it does not go to the correct LOD pair unless you descend or climb a little to trigger the recalculation. It also resolves the issue you stated where it is skipping pairs and not being in the right LOD pair at cruise because it is now active at cruise and will fix that right away.

If you are still worried about LODs toggling too often in cruise then just make the null zone higher than the 5% I chose. eg. if you make it 20% and you are cruising at 30000 ft and have a LOD trigger at 30000 ft, then it will only change if AGL changes by 6000 ft which would be extremely rare for all those conditions to exist at the same time. ie. for the most part in airliner cruising, this code lays dormant.

**Comment by ResetXPDR** [comment]
Just to illustrate how this change minimally impacts airliner ops, attached is a log file of an airliner flight I made today over north east USA and at no time during cruise (between 09:20:53.639 and 09:53:42.257) did my code changes activate. 

[DynamicLOD20240115.log](https://github.com/Fragtality/DynamicLOD/files/13933634/DynamicLOD20240115.log)

**Comment by Fragtality** [comment]
Was that a Step Descent?
![image](https://github.com/Fragtality/DynamicLOD/assets/78614902/aeaa6864-fb27-4cc9-a197-f5cf0b2a7359)

Constantly writing to Memory, not Good.
![image](https://github.com/Fragtality/DynamicLOD/assets/78614902/603c9f23-63c3-4647-bdde-9b6e4c15d1dd)

**Comment by ResetXPDR** [comment]
Descent was normal, not stepped.

Those OLOD 30 lines do not come from my code changes as log entries from my code do not start with LODController:RunTik. They were in the climb phase or descent phase, where my code is not active.

**Comment by ResetXPDR** [comment]
Ah, I know what those OLOD 30 lines are from. When I am flying in VR and take the headset off, I Iimit FPS to 20 to save power and noise. For some reason that runtick code is triggering despite me not having FPS adaption selected as shown which is a bit strange, but not from my code changes.
![image](https://github.com/Fragtality/DynamicLOD/assets/156568588/821ea1db-c9fe-4b25-bcd3-22dd68f3ba98)

**Comment by ResetXPDR** [comment]
Here's my log from 2 days ago before I made any changes to the code and it has quite a few runtick OLOD 30 lines too.
[DynamicLOD20240113.log](https://github.com/Fragtality/DynamicLOD/files/13933757/DynamicLOD20240113.log)

**Comment by ResetXPDR** [comment]
I'll try and quantify what is triggering the OLOD 30 issue and will get back to you.

**Comment by Fragtality** [comment]
Sorry if the Exception in actually answering an Issue made a an false Impression ... I'm not really interested in "will get back to you".

**Comment by ResetXPDR** [comment]
I'm not really sure why you are getting offended. I just answered your questions with facts. I'm sorry you feel that way.

BTW, the OLOD 30 issue is happening in the climb in a test I am doing right now, even without FPS cap on.
![image](https://github.com/Fragtality/DynamicLOD/assets/156568588/06e362ba-a650-43d5-987b-2ec59d5cef19)

**Comment by ResetXPDR** [comment]
I found what is causing the RunTick OLOD 30 line to keep occurring. For some reason when my aircraft is above 5000 ft but below 10000 ft, so is using the OLOD 30 at 5000 ft line, the olod value being compared on line 110 is actually 30.000002 against newlod value of 30, hence is not equal and is triggering the code to set OLOD to 30. All of the other OLOD values (100, 70, 50 and 10) show as whole numbers but 30 is being read as 30.000002, even when I reentered the data in the app for that level.
As a quick fix I modified the line 110 to be: 
if (Math.Round (olod) != newlod)
and the erroneous log entry and OLOD setting goes away. While this fixes it, I'd be more worried what is setting olod to 30.000002 in the first place as that is not the value in the table.
Anyway, FYI.

Edit: looks like olod is what is being read in from MSFS memory, not from the table, so not much you can do about it other than round it like I did for the comparison (same with tlod for completeness).

With olod and tlod rounded as suggested, my last flight went perfectly with no erroneous OLOD 30 entries. 
![image](https://github.com/Fragtality/DynamicLOD/assets/156568588/4217e961-7a75-41b0-a9dd-c6593dbccdac)

It's probably better to round olod and tlod at the point of reading from MSFS memory in the GetXLOD functions in the MemoryManager class so that other uses of these variables don't fall into the same trap. eg. 
![image](https://github.com/Fragtality/DynamicLOD/assets/156568588/2fc17aff-c6df-483c-be90-0b57ee7c9ead)
plus the other three functions.

**Comment by Fragtality** [comment]
> I'm not really sure why you are getting offended. I just answered your questions with facts. I'm sorry you feel that way.

Uhm - I'm not getting offended, I just lost interest in the Discussion. Being disinterested is not the same as being offended?
<br/>


I mean honestly, what exactly did you expect? To summarize: You have the Idea of making a Suggestion to a Tool where the very Description of that Tool already states that there is no Interest in Suggestions. Then you go ahead and formulate your Suggestion to the wrong Project/Repo first. Then the Suggestions is something to work-around the Tool's Core-Principle. A Suggestion based on a wrong Assumption (to quote: "... will attempt to modify it to continuously select LOD based on AGL if in cruise mode ..."). And which even actually enables said Behavior.
So your initial Conditions could not have been worse.

So in that Sense you where quite successful imho, since the standard Reaction to any Issue here is an instant-Close (sometimes not even with a Comment). But you certainly put considerable Effort in your Suggestion, so I don't wanted it to answer in that Way - albeit the Fact that it was getting late here and I was still working on other Stuff I just care more about.

And certainly it wasn't all in vain. Your Findings confirm my Idea that there needs to be some kind of an additional Checks/Cases (if it actually should attempt to change the LOD based on the AGL when in Cruise). And the Finding that we actually can have Precision Errors here is really really good. (Which is possibly a Regression of Simplification ...)

But what is also certain: I won't accept your (initial) proposed Changes. I won't give up on the Principle that the vertical Behavior of the Airplane has to be taken into Consideration.
After all it is an Open-Source Project under MIT License. So if you think that another Approach is more reasonable, there is nothing stopping you from doing a Fork, making the Changes and publishing that on your forked Repo. Heck, you can even do more welcoming Description and answer Issues & Suggestions of Users.
